### ISSUE TITLE
getters, setters and Meta Properties

### ISSUE THREAD

**Issue by dmill-bz** [spider/spider#12]
recap : 

---

Also another issue I noticed is that graph elements have metaProperties and properties. 
Meta properties are things like `id`, `label`, `type`. Properties on the other hand can be anything the user sets (including one of those three). With the current structure where SpiderGraph is populated with a combination of both this can lead to conflicts (if say the vertex has a `type` property for example).

Consider the following to see where conflicts could arise (you can see that properties is set in SpiderRecord to avoid said conflicts):

``` php
    /**
     * Hydrate a SpiderRecord from a gremlin-server response array
     */
    protected function arrayToSpiderRecord($array)
    {
        // Or we map a single record to a Spider Record
        $spiderRecord = new SpiderRecord();
        $properties = [];
        foreach($array['properties'] as $key => $value)
        {
            $properties[$key] = $value[0]['value'];
        }

        $spiderRecord->add([
            'id' => $array['id'],
            'label' => $array['label'],
            'properties' => $properties,
        ]);
        return $spiderRecord;
    }
```

---

**Meta Properties**

Completely agree. Maybe records should be mapped in such a way that every result has a meta bag with meta properties?

``` php
$result = query->one();
$result->username; // normal property for username
$result->birthday; // normal property birthday
$result->label; // normal property for label, maybe like a tag system or whatever
$result->meta->label; // meta property for class or type, automatically populated
```

While on that topic. We should come up with a standard vocabulary. Different databases have different terms for label, class, property, field, etc. May be best just to stick with something like Neo4j or Gremlin since they are the most used.

**Comment by dmill-bz** [comment]
I think we should base ourselves on TinkerPop here as it's a collection of all possible combinations. 

A few things that might need thinking if we go the `meta` route:
- is `$record->meta->id` acceptable to get the record id? Seems weird.
- how does one extend `Record` to make a `UserRecord` with a `label = "user"`
- not all databases accept custom id. Do we need to make sure users don't override this property?

Some getter/setter thoughts:

``` php
$record = new Record(); // empty record
$record->meta->label = "user";
$record->name = "john"; //does this work? What are the dangers?
$record->age = 25;
$record->getAge(); // returns 25, does this work out of the box?
$record->save(); // does a $record->insert() since no ID is set.
$record->meta->id; // some id

$newRecord = Record::select()->where('name', 'john')->one();
$newRecord->name; // john
$newRecord->age; // 25
$newRecord->meta->id; // some id (same as previous)
```

**Comment by electricjones** [comment]
I have been considering a `Record` to be a plain-php-object with properties from the database. Maybe a helper method or two. I had not considered `Records` to be connected to the datastore. That is, once a record is sent back from the driver, it is finished. You must execute a new query to get anything else form the driver.

What you describe above, I had called Models which are extensions of records. That is, records that maintain a link to the driver to expand themselves or whatever. Laravel makes that distinction in their ORM. When using the `QueryBuilder` to return data, a `Collection` is returned, which is nothing more than an array on steroids.

``` php
// From laravel
$result = Query::table('users')->where('username', 'michael')->get();
$result->name; // first user's name.
// $result also has a bunch of helper methods like pop() and count()

// But you cannot
$result->name = "jon"
$result->save();

// You would have to
Query::table('users')->update($result->id)->set('name', jon');
```

When using the Models, you have the above functionality.

``` php
$user = User::where('username', 'michael');
$user->name = "jon"
$user->save();
```

I had planned on using the same methodology.

**Comment by electricjones** [comment]
What if we included at the top level `$id` and `$_label` and then a meta object to hold any other metadata. For instance, OrientDB includes cluster and position information. ID and label are small enough, that if they were in both the top-level record and the meta object, that would be fine.

It is common practice to simply preface a protected variable with an "_". Would this be enough?

So, yes, a `UserRecord` (though I would call that a Model) would have:

``` php
$user->_label; // user
$user->id; //it may be of note that ids work differently. Orient ids are strings with cluster and position (#14:1234)

$user->name; // regular properties
$user->birthday;

$user->meta->cluster; // specific to datastore
```

**Comment by dmill-bz** [comment]
ok SO... I realize from a couple of our discussions that I've considered `Record` to be a `Model`. Hence a big part of the confusion. Personal preference would go towards naming it `Collection` or `Response extends Collection` 

This makes a lot of sense.
I also agree with keeping `id` and `label` directly in class. Not much more is required really.
I like where this is going.

**Comment by electricjones** [comment]
I posted a long response on #11, but to wrap up this as well. I like using consistent getters and setters. I would recommend basing any object that is hydrated on the fly off of `michaels/data-manager` which is already used in production and required by spider. I can also make any changes needed, since its my repo.

I can add some protection features to data-manager that allows for making items immutable. Data-manager uses traits or extending classes. It can be extended with helpers for collections. With it, you can

``` php
$manager = new Michaels\Manager\Manager([
    'some' => [
        'starting' => [
            'data' => 'here (optional)'
        ]
    ]
]);

/* Basic Usage. All works with dot notation as well */
$manager->add('name', 'value');
$manager->add('some.nested.data', 3); // Use dot notation for namespacing or nesting
$manager->get('name'); // 'value'
$manager->get('doesntexist', 'fallback'); // 'fallback'
$manager->all(); // returns array of all items
$manager->exists('name'); // true
$manager->set('name', 'new-value'); // updates item
$manager->remove('some.starting.data');
$manager->toJson(); // returns json of all items

/* You can also use $manager as an array or in loops */

/* Finally, you may access values using magic methods */
$manager->one;
$manager->one = 'new one';
$manager->some()->starting()->data; //nesting
```

**Comment by dmill-bz** [comment]
Yeah this looks good. I'm sure it'll unlock quite a few things in the future as well!

**Comment by electricjones** [comment]
I can work through some of the value objects tomorrow and get this done and closed

**Comment by electricjones** [comment]
I will close this after the #20 merge
